cache_ui/self_monitor.go

```diff
@@ -0,0 +1,236 @@
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	selfTestBody string = "This object was created by the Pelican self-test functionality"
+	selfTestDir  string = "/pelican/monitoring"
+)
+
+func InitSelfTestDir() error {
+	uid, err := config.GetDaemonUID()
+	if err != nil {
+		return err
+	}
+
+	gid, err := config.GetDaemonGID()
+	if err != nil {
+		return err
+	}
+
+	basePath := param.Cache_DataLocation.GetString()
+	pelicanMonPath := path.Join(basePath, "/pelican")
+	monitoringPath := path.Join(pelicanMonPath, "/monitoring")
+	err = os.MkdirAll(monitoringPath, 0700)
+	if err != nil {
+		return errors.Wrap(err, "Fail to create directory for the self-test")
+	}
+	if err = os.Chown(pelicanMonPath, uid, gid); err != nil {
+		return errors.Wrapf(err, "Unable to change ownership of self-test /pelican directory %v to desired daemon gid %v", monitoringPath, gid)
+	}
+	if err = os.Chown(monitoringPath, uid, gid); err != nil {
+		return errors.Wrapf(err, "Unable to change ownership of self-test /pelican/monitoring directory %v to desired daemon gid %v", monitoringPath, gid)
+	}
+	return nil
+}
+
+func generateTestFile() (string, error) {
+	basePath := param.Cache_DataLocation.GetString()
+	monitoringPath := path.Join(basePath, selfTestDir)
+	_, err := os.Stat(monitoringPath)
+	if err != nil {
+		return "", errors.Wrap(err, "Self-test directory does not exist")
+	}
+	uid, err := config.GetDaemonUID()
+	if err != nil {
+		return "", err
+	}
+
+	gid, err := config.GetDaemonGID()
+	if err != nil {
+		return "", err
+	}
+
+	now := time.Now()
+	testFileBytes := []byte(selfTestBody)
+	fileSize := len(testFileBytes)
+	cinfo := CInfo{
+		Store: Store{
+			FileSize:     int64(fileSize),
+			CreationTime: now.Unix(),
+			Status:       2, // CSChk_None = 0
+		},
+	}
+	cinfoBytes, err := cinfo.Serialize()
+	if err != nil {
+		return "", err
+	}
+	testFileName := "self-test-" + now.Format(time.RFC3339) + ".txt"
+	testFileCinfoName := "self-test-" + now.Format(time.RFC3339) + ".txt.cinfo"
+
+	filePath := path.Join(monitoringPath, testFileName)
+	extFilePath := path.Join(selfTestDir, testFileName)
+	fileCinfoPath := path.Join(monitoringPath, testFileCinfoName)
+
+	file, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to create self-test file %s", filePath)
+	}
+	defer file.Close()
+
+	cinfoFile, err := os.OpenFile(fileCinfoPath, os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to create self-test cinfo file %s", fileCinfoPath)
+	}
+	defer cinfoFile.Close()
+
+	if _, err := file.Write(testFileBytes); err != nil {
+		return "", errors.Wrapf(err, "Failed to write test content to self-test file %s", filePath)
+	}
+
+	if _, err := cinfoFile.Write(cinfoBytes); err != nil {
+		return "", errors.Wrapf(err, "Failed to write cinfo content to self-test cinfo file %s", fileCinfoPath)
+	}
+
+	if err = os.Chown(filePath, uid, gid); err != nil {
```

Comment:
Use [File.Chown](https://pkg.go.dev/os#File.Chown) instead.  That will chown the open file handle.